kernel/generic/src/synch/syswaitq.c

```diff
@@ -0,0 +1,189 @@
+/*
+ * Copyright (c) 2018 Jakub Jermar
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ *
+ * - Redistributions of source code must retain the above copyright
+ *   notice, this list of conditions and the following disclaimer.
+ * - Redistributions in binary form must reproduce the above copyright
+ *   notice, this list of conditions and the following disclaimer in the
+ *   documentation and/or other materials provided with the distribution.
+ * - The name of the author may not be used to endorse or promote products
+ *   derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/** @addtogroup kernel_sync
+ * @{
+ */
+
+/**
+ * @file
+ * @brief Wrapper for using wait queue as a kobject.
+ */
+
+#include <synch/syswaitq.h>
+#include <synch/waitq.h>
+#include <abi/cap.h>
+#include <cap/cap.h>
+#include <mm/slab.h>
+#include <proc/task.h>
+#include <syscall/copy.h>
+
+#include <stdint.h>
+
+static slab_cache_t *waitq_cache;
+
+static void waitq_destroy(void *arg)
+{
+	waitq_t *wq = (waitq_t *) arg;
+	slab_free(waitq_cache, wq);
+}
+
+static kobject_ops_t waitq_kobject_ops = {
+	.destroy = waitq_destroy
+};
+
+static bool waitq_cap_cleanup_cb(cap_t *cap, void *arg)
+{
+	kobject_t *kobj = cap_unpublish(cap->task, cap->handle,
+	    KOBJECT_TYPE_WAITQ);
+	kobject_put(kobj);
+	cap_free(cap->task, cap->handle);
+	return true;
+}
+
+/** Initialize the user waitq subsystem */
+void sys_waitq_init(void)
+{
+	waitq_cache = slab_cache_create("waitq_t", sizeof(waitq_t), 0, NULL,
+	    NULL, 0);
+}
+
+/** Clean-up all waitq capabilities held by the exiting task */
+void sys_waitq_task_cleanup(void)
```

Comment:
So, why do we need a separate cleanup routine for each kobject type, instead of just one generic cleanup?

Reply:
Slightly different semantics of the objects. Sometimes you need to do further actions. This should eventually become normalized, but not now.